docs/source/miscellaneous/doc/adr/0013-write-cassandra-tombstones-and-sentinels-with-a-fresh-cassandra-timestamp.md

```diff
@@ -0,0 +1,109 @@
+# 13. Write Cassandra tombstones and sentinels with a fresh Cassandra timestamp
+
+Date: 12/06/2018
+
+## Status
+
+Accepted
+
+## Context
+
+### Mutations
+
+Mutations in Cassandra take place at a Cassandra *writetime* or timestamp. In normal Cassandra operation, these times
+are the wall-clock times where a mutation took place. These are used to "resolve" conflicts when performing reads;
+for a given column, the most recently written value will be read ("last-write-wins").
+
+AtlasDB columns take the form `(row, column1, column2, value)` where `column2` is the Atlas timestamp. These timestamps
+are unique - thus, for transactional tables, no mutations to the same column should take place apart from removing
+a cell that is no longer needed (e.g. because of sweep).
+
+In the context of AtlasDB, reliance on wall-clock time is generally considered unacceptable. Thus, Atlas uses its
+own timestamps:
+
+- a cell written at a start timestamp `TS` is given the Cassandra timestamp `TS`
+- a sweep sentinel is given the Cassandra timestamp `-1`
+- a tombstone inserted by deleting a cell with start timestamp `TS` is given the Cassandra timestamp `TS + 1`
+- a range tombstone from time `0` to `TS` exclusive is given the Cassandra timestamp `TS`
+
+For transactional tables, there are several key invariants to be preserved:
+
+- a tombstone or range tombstone that covers a cell with the Cassandra timestamp `TS` must have a Cassandra timestamp 
+  greater than `TS`
+- a tombstone or range tombstone that covers a sweep sentinel with Cassandra timestamp `TS` must have a Cassandra
+  timestamp greater than `TS`
+- the insertion of a fresh sweep sentinel after a tombstone or range tombstone at time `TS` must have a Cassandra
```

Comment:
I'm not sure this is correct? It depends on what the range of the tombstone is no? We write sentinel at `column2 = -1` so as long as that isn't in the range we're fine?

Reply:
Yeah that's right - we only care if the tombstone covers the sentinel. It should be `the insertion of a fresh sweep sentinel after a (possibly range) tombstone that covers it that was written at TS must have a C* timestamp greater than TS`